s02/2.md

```diff
@@ -10,14 +10,556 @@ Tartalom
+
+Továbbra sem ajánlott `new`-val kivételt létrehozni. Ha a dianamikus típussal akarunk játszani,
+inkább használjuk a következő megoldást:
+
+```C++
+struct ExceptionBase{
+    virtual void raise() { throw *this; }
+    virtual ~ExceptionBase() {}
+};
+
+struct ExceptionDerived : ExceptionBase{
+    virtual void raise() { throw *this; }
+};
+
+void foo(ExceptionBase& e){
+    e.raise(); // Uses dynamic type of e while raising an exception.
+}
+int main (void){
+    ExceptionDerived e;
+    try {
+        foo(e);
+    }catch (ExceptionDerived& e) {
+        ...
+    }catch (...) {
+        ...
+    }
+}
+```
+
+#### Kivételkezelés és osztályok
+
+Kérdéses esetek: konstruktor, destruktor
+
+* Konstruktor
+    + Ha a konstruktor dob, terület le lett foglalva, de a pointer nem lett beállítva $\Rightarrow$ nincs gond,
+    a rendszer deallokálja
+    + De: a konstruktoron belül lefogalt területet a destruktor tudja felszabadítani. A destruktort viszont nem lehet 
+    meghívni mert az objektum létre se jött rendesen $\Rightarrow$ probléma
+    ```C++
+    class X
+    {
+        public:
+            X(int i) { p = new char[i]; init(); }
+            ~X() { delete [] p; }       // must not throw exception
+        private:
+            void init() { ... throw ... }   // BAD: destructor won't run !
+            char *p;                        // constructor was not completed
+    };
+    ```
+    
+    + Ha tagváltozó inicializálása dob hibát, akkor dob a konstruktor is
+    ```C++
+    class X
+    {
+        public:
+            X() { throw 1; }
+    };
+    class Y
+    {
+        public:
+            Y()
+            try
+                : x()
+            { }
+            catch( ... ) { /* throw; */ }
+        private:
+            X x;
+    };
+    int main(){
+        try {
+            Y y;
+            return 0;
+        }
+        catch (int i)
+        {
+            std::cerr << "exception: " << i << std::endl;
+        }
+    }
+    ```
+    
+* Destruktor
+    + destruktorokat kétféle okból hívunk
+        - nomális esetben
+        - kivételkezeléskor $\Rightarrow$ ha a destruktor is kivételt dob az nem definiált
+        viselkedéshez vezet (legtöbbször a `terminate()` meghívásához)
+    + $\Rightarrow$ Ökölszabály: **destruktor nem dobhat kivételt**
+
+#### Noexcept (C++11)
+
+* Kifejezhető vele, hogy egy kifejezés, függvény biztosan nem dob-e exceptiont
+* Fordítási időben értékelődik ki
+
+Kétféle formában létezik:
+
+* operátor forma: `bool noexcept(expr);`
+    + Nem értékeli ki a kifejezést (hasonló a `sizeof`-hoz)
+    + `false`, ha
+        - a kifejesés dob
+        - a kifejesésben van dynamic_cast
+        - a kifejesésnek van `type_id`-ja
+        - van a kifejesésben függvény, ami nem `noexcept(true)` és nem `constexpr`
+    + `true` különben
+
+* specifier forma: `void f() noexcept(expr) { }`
+    + a régi `throw()` helyett van
+    + pl.: Ha `g()` nem dob, akkor `f` sem:
+    ```C++
+        template <typename T>
```

Comment:
Azt akartam elérni, hogy be legyen húzva, mert ugye a bullet point-hoz tartozik. Hülyén nézett ki, hogy a felsorolás beljebb van kezdve, a kódpélda meg teljesen a szélén.